### PR TITLE
Update rich workspace visibility

### PR DESCRIPTION
**Pull request by luka-nextcloud**
Signed-off-by: Luka Trovic <luka@nextcloud.com>

* Depends on https://github.com/nextcloud/server/pull/35474
* Resolves: #3209
* Target version: master

### CODE REVIEW

File: src/views/RichWorkspace.vue [hunk 1]
```diff
@@ -22,17 +22,6 @@
 
 <template>
 	<div v-if="enabled" id="rich-workspace" :class="{'icon-loading': !loaded || !ready, 'focus': focus, 'dark': darkTheme, 'creatable': canCreate, 'empty': showEmptyWorkspace}">
```
Comment (juliushaertl) [inline src/views/RichWorkspace.vue:24]: We probably don't need the empty class anymore and can drop showEmptyWorkspace

File: src/views/RichWorkspace.vue [hunk 2]
```diff
@@ -221,6 +190,72 @@ export default {
...
+											.createFile('Readme.md', { scrollTo: false, animate: false })
+											.then(() => {
+												menu.removeMenuEntry('rich-workspace-init')
+												return self.getFileInfo()
+											})
+									},
+								})
+							},
+						}
+						OC.Plugins.register('OCA.Files.NewFileMenu', newRichWorkspaceFileMenuPlugin)
```
Comment (juliushaertl) [inline src/views/RichWorkspace.vue:249]: I think this does not work when changing directories as the plugin would be registered each time or the folder info would only be fetched for the first one. 

I think the best would be to move this to https://github.com/nextcloud/text/blob/81d7af739e8c365793888aaa40105ac03f79843e/src/files.js#L60

Comment (juliushaertl) [reply src/views/RichWorkspace.vue:249]: Show dynamically showing or hiding we can probably use the same events as in https://github.com/nextcloud/text/blob/e286194c690c15c5d66b3a7198ee76b9366f1201/src/helpers/files.js#L177-L182

Comment (khoipv-1315) [reply src/views/RichWorkspace.vue:249]: I have a question, how can I hide the menu?
- If using Document Select, it won't find the menu because it doesn't show the menu yet if it doesn't click the menu button
- Is there a way to add a Menu item using FileList? I only know how to use OCA.Files.NewFileMenu

File: src/helpers/files.js [hunk 3]
```diff
@@ -139,14 +139,48 @@ const registerFileActionFallback = () => {
...
+			fileType: 'file',
+			actionHandler() {
+				return window.FileList
+					.createFile('Readme.md', { scrollTo: false, animate: false })
+					.then(() => {
+						menu.removeMenuEntry('rich-workspace-init')
+					})
+			},
+			shouldShow() {
+				const descriptionFile = 'Readme.md'
```
Comment (juliushaertl) [inline src/helpers/files.js:165]:
```suggestion
				const descriptionFile = t('text', 'Readme') + '.' + loadState('text', 'default_file_extension'),
```

File: src/helpers/files.js [hunk 4]
```diff
@@ -139,14 +139,48 @@ const registerFileActionFallback = () => {
...
+		if (fileList.id !== 'files' && fileList.id !== 'files.public') {
+			return
+		}
+
+		// register the new menu entry
+		menu.addMenuEntry({
+			id: 'rich-workspace-init',
+			displayName: t('text', 'Add description'),
+			templateName: t('text', 'Readme') + '.' + loadState('text', 'default_file_extension'),
+			iconClass: 'icon-filetype-text',
```
Comment (juliushaertl) [inline src/helpers/files.js:155]:
```suggestion
			iconClass: 'icon-rename',
```

To match the design mockup and have distinguishable icons:

Screenshot 2022-12-08 at 14 27 06

File: src/helpers/files.js [hunk 5]
```diff
@@ -139,14 +139,48 @@ const registerFileActionFallback = () => {
 
 }
 
+const newRichWorkspaceFileMenuPlugin = {
+	attach(menu) {
+		const fileList = menu.fileList
+		// only attach to main file list, public view is not supported yet
+		if (fileList.id !== 'files' && fileList.id !== 'files.public') {
```
Comment (juliushaertl) [inline src/helpers/files.js:146]: Public file lists could still be supported. I think we also cover that with out existing implementation, but needs double checking.

File: src/helpers/files.js [hunk 6]
```diff
@@ -171,22 +206,24 @@ const FilesWorkspacePlugin = {
...
 				vm.active = OCA.Files.App.getCurrentFileList() === fileList
 			})
-
 			fileList.$el.on('urlChanged', data => {
 				vm.path = data.dir.toString()
 			})
 			fileList.$el.on('changeDirectory', data => {
 				vm.path = data.dir.toString()
 			})
+			fileList.$table.on('DOMSubtreeModified', () => {
```
Comment (juliushaertl) [inline src/helpers/files.js:224]:
```suggestion
			fileList.$table.on('afterChangeDirectory', () => {
```

I think this should also work and is only triggered when the list of files is actually updated

File: src/helpers/files.js [hunk 7]
```diff
@@ -160,9 +194,10 @@ const FilesWorkspacePlugin = {
 		if (fileList.id !== 'files' && fileList.id !== 'files.public') {
 			return
 		}
-
+		addMenuRichWorkspace()
```
Comment (juliushaertl) [inline src/helpers/files.js:197]: I'd say we can just move the `menu.addMenuEntry` call in here and avoid the two plugins being nested to just have one for the rich workspace.

File: src/helpers/files.js [hunk 8]
```diff
@@ -171,22 +206,24 @@ const FilesWorkspacePlugin = {
...
 			})
-
 			fileList.$el.on('urlChanged', data => {
 				vm.path = data.dir.toString()
 			})
 			fileList.$el.on('changeDirectory', data => {
 				vm.path = data.dir.toString()
 			})
+			fileList.$table.on('DOMSubtreeModified', () => {
+				vm.hasDescriptionFile = !!fileList.findFile(descriptionFile)
```
Comment (juliushaertl) [inline src/helpers/files.js:225]: I'd prefer to have an event that is triggered through https://github.com/nextcloud/nextcloud-event-bus on the newFileAction after creating the file. The RIchWorkspace.vue component could listen to that and reuse the existing way to check for a readme file through the backend https://github.com/nextcloud/text/blob/master/src/views/RichWorkspace.vue#LL145

File: src/views/RichWorkspace.vue [hunk 9]
```diff
@@ -256,42 +225,42 @@ export default {
 		color: var(--color-text-maxcontrast);
 	}
 
-	#rich-workspace:deep(div[contenteditable=false]){
+	#rich-workspace div[contenteditable=false] {
```
Comment (juliushaertl) [inline src/views/RichWorkspace.vue:228]: What is the reason for that change? Should not be needed

File: src/helpers/files.js (hunk 5 above)
Comment (luka-nextcloud) [reply src/helpers/files.js:146]: How do you suggest that we proceed here?

File: src/helpers/files.js (hunk 6 above)
Comment (luka-nextcloud) [reply src/helpers/files.js:224]: This event is not triggered when we add the README.md file to the home directory. Do you have any alternative way?

File: src/helpers/files.js (hunk 5 above)
Comment (juliushaertl) [reply src/helpers/files.js:146]: Noticed that we also exclude this in the existing plugin so let's skip that for now.

File: src/helpers/files.js (hunk 6 above)
Comment (juliushaertl) [reply src/helpers/files.js:224]: That should be fine if we rely on the existing mechanism to check for a readme existence in the component as noted below https://github.com/nextcloud/text/pull/3291#discussion_r1043361054

File: src/helpers/files.js (hunk 8 above)
Comment (luka-nextcloud) [reply src/helpers/files.js:225]: I definitely agree with that. There is a problem though. When I tried, I noticed that the UI takes a long time before updating the rich workspace block. I guess because of the event-bus.

Comment (juliushaertl) [reply src/helpers/files.js:225]: It should take the same amount of time as it is currently. There is an extra request but that should not change from how it is right now. Event bus in general should be fast ;)

File: src/views/RichWorkspace.vue (hunk 9 above)
Comment (luka-nextcloud) [reply src/views/RichWorkspace.vue:228]: I noticed that they were not applied to the editor and I was not able to work with the editor normally.

File: src/helpers/files.js (hunk 8 above)
Comment (luka-nextcloud) [reply src/helpers/files.js:225]: The problem is `DOMSubtreeModified` is triggered almost all the time so it's always triggering a new event (either `Text::showRichWorkspace` or `Text::hideRichWorkspace`)

Comment (juliushaertl) [reply src/helpers/files.js]: Yes, I think that is triggered way too often, therefore I'd go with the above mentioned event bus approach.

File: src/views/RichWorkspace.vue (hunk 9 above)
Comment (juliushaertl) [reply src/views/RichWorkspace.vue:222]: Can you maybe check if that works after running `npm ci` I think this might only be the case if you have some outdated dependencies.

Comment (juliushaertl) [reply src/views/RichWorkspace.vue:222]: If that is still an issue we should still address it in a separate PR I think.

Comment (luka-nextcloud) [reply src/views/RichWorkspace.vue:222]: Ok, let me check after npm ci.

File: src/helpers/files.js (hunk 8 above)
Comment (luka-nextcloud) [reply src/helpers/files.js]: The event bus approach is already implemented. The actual problem is the right event to use. We're using `DOMSubtreeModified` right now.

File: src/helpers/files.js [hunk 10]
```diff
@@ -139,14 +139,46 @@ const registerFileActionFallback = () => {
...
+		}
+
+		// register the new menu entry
+		menu.addMenuEntry({
+			id: 'rich-workspace-init',
+			displayName: t('text', 'Add description'),
+			templateName: descriptionFile,
+			iconClass: 'icon-rename',
+			fileType: 'file',
+			useInput: descriptionFile,
```
Comment (juliushaertl) [inline src/helpers/files.js:158]:
```suggestion
			useInput: false,
```

To make sure that the input is not shown

File: src/helpers/files.js (hunk 8 above)
Comment (luka-nextcloud) [reply src/helpers/files.js]: @juliushaertl Is there any event we can listen to when the README.md file is deleted?

File: src/views/RichWorkspace.vue (hunk 9 above)
Comment (luka-nextcloud) [reply src/views/RichWorkspace.vue]: @juliushaertl It actually doesn't work.

File: src/helpers/files.js (hunk 8 above)
Comment (luka-nextcloud) [reply src/helpers/files.js]: @juliushaertl Could you please check again?

File: src/views/RichWorkspace.vue (hunk 9 above)
Comment (luka-nextcloud) [reply src/views/RichWorkspace.vue]: > If that is still an issue we should still address it in a separate PR I think.

I will remove the changes from the current PR and open a new one for the fix.

Comment (luka-nextcloud) [reply src/views/RichWorkspace.vue:222]: Removed the changes for now.

File: src/helpers/files.js (hunk 8 above)
Comment (juliushaertl) [reply src/helpers/files.js]: Nothing yet. We could skip this for now as I know there is some ongoing work in the files2vue migration.

Otherwise a server PR with triggering an event around here would be the way to go I think:

https://github.com/nextcloud/server/blob/5b9a8f0407a9e3fe7e00f0fc9284ea986905f1b5/apps/files/js/filelist.js#L2494

Comment (juliushaertl) [reply src/helpers/files.js]: Something like: `this.$el.trigger(jQuery.Event('removeFile', path));`

**Comment by juliushaertl** [comment]
@luka-nextcloud Server pr was merged, so rebasing should run cypress against the latest server master branch.

File: cypress/e2e/workspace.spec.js [hunk 11]
```diff
@@ -62,27 +62,33 @@ describe('Workspace', function() {
...
 			.should('be.visible')
-			.should('contain', 'Hello')
 	})
 
 	it('formats text', function() {
 		cy.visit(`apps/files?dir=/${encodeURIComponent(currentFolder)}`)
 		cy.openWorkspace()
-			.type('Format me')
-			.type('{selectall}')
+			.type('Format me', { force: true })
```
Comment (juliushaertl) [inline cypress/e2e/workspace.spec.js:90]: Would be good to have a way without force. Maybe we need to focus the workspace and wait for the menubar to appear first.

File: cypress/e2e/workspace.spec.js [hunk 12]
```diff
@@ -104,8 +110,8 @@ describe('Workspace', function() {
 	it('creates headings via submenu', function() {
 		cy.visit(`apps/files?dir=/${encodeURIComponent(currentFolder)}`)
 		cy.openWorkspace()
-			.type('Heading')
-			.type('{selectall}')
+			.type('Heading', { force: true })
```
Comment (juliushaertl) [inline cypress/e2e/workspace.spec.js:113]: Same as above.

File: cypress/support/commands.js [hunk 13]
```diff
@@ -301,7 +301,7 @@ Cypress.Commands.add('getMenuEntry', (name) => {
 Cypress.Commands.add('getSubmenuEntry', { prevSubject: 'optional' }, (subject, parent, name) => {
 	return (subject ? cy.wrap(subject) : cy.getMenu())
 		.getActionEntry(parent)
-		.click()
+		.click({force: true})
```
Comment (juliushaertl) [inline cypress/support/commands.js:304]: Seems unrelated

File: cypress/support/commands.js [hunk 14]
```diff
@@ -337,8 +337,9 @@ Cypress.Commands.add('clearContent', () => {
 })
 
 Cypress.Commands.add('openWorkspace', () => {
-	cy.get('#rich-workspace .empty-workspace').click()
-	cy.getEditor().find('[data-text-el="editor-content-wrapper"]').click()
+	cy.createDescription()
+	cy.get('#rich-workspace .editor__content').click({force: true})
```
Comment (juliushaertl) [inline cypress/support/commands.js:341]: This looks like the right place to wait until the editor content is visible. instead of forcing actions all over the place ;)

File: src/helpers/files.js [hunk 15]
```diff
@@ -139,14 +139,47 @@ const registerFileActionFallback = () => {
...
+			fileType: 'file',
+			useInput: false,
+			actionHandler() {
+				return window.FileList
+					.createFile(descriptionFile, { scrollTo: false, animate: false })
+					.then(() => emit('Text::showRichWorkspace'))
+			},
+			shouldShow() {
+				if (fileList.findFile(descriptionFile)) return false
+				return true
```
Comment (juliushaertl) [inline src/helpers/files.js:166]:
```suggestion
				return !fileList.findFile(descriptionFile)
```

File: src/helpers/files.js [hunk 16]
```diff
@@ -139,14 +139,47 @@ const registerFileActionFallback = () => {
 
 }
 
+const newRichWorkspaceFileMenuPlugin = {
+	attach(menu) {
```
Comment (juliushaertl) [inline src/helpers/files.js:143]: Can we merge this into the existing FilesWorkspacePlugin below? There should be no need for two separate plugins

**Review by juliushaertl — CHANGES_REQUESTED** [review]

**Comment by juliushaertl** [comment]
@luka-nextcloud Cypress tests seem good now, can you remove the remaining force parts in the spec files and fix the eslint errors?

File: src/helpers/files.js (hunk 16 above)
Comment (luka-nextcloud) [reply src/helpers/files.js:143]: The plugin below gets the FileList as argument for the attach method but we need the menu instead.

Comment (juliushaertl) [reply src/helpers/files.js:143]: Right ... sorry I mixed that up then 👍

**Review by juliushaertl — APPROVED** [review]
